### PR TITLE
llm-proxy: implement initial rate-limiting and rate limit updates

### PR DESCRIPTION
**Pull request by bobheadxi**
Initial direction:

- Product subscriptions get updated opportunistically in hot-path of request if:
  - the record is old
  - a request hits a rate limit, and the record hasn't been updated recently
- Actors are the entity for rate-limiting, retrieved from actor.Source implementations that could later be another source or simply a static IP-based limiter

This aligns with [this model](https://stately.ai/registry/editor/f0eed212-1730-4052-abd7-b36689e0f50f?machineId=8cbb3591-a750-486d-a97e-cef2dc2b90e7&mode=Simulate) BUT I think we have consensus that a dumber full-sync model is probably easier to reason with, so will be implementing that soon(tm).

## Test plan

Manual testing only because we are likely replacing this system soon, not a critical dependency yet

```
you exceeded the rate limit for completions. Current usage: 50 out of 50 requests. Retry after 2023-05-06 17:04:12 +0000 UTC
```

### CODE REVIEW

File: enterprise/cmd/llm-proxy/internal/actor/actor.go [hunk 1]
```diff
@@ -56,7 +54,47 @@ func FromContext(ctx context.Context) *Actor {
...
+}
+
+func (u updateOnFailureLimiter) TryAcquire(ctx context.Context) error {
+	err := (limiter.StaticLimiter{
+		Identifier: u.ID,
+		Redis:      u.Redis,
+		Limit:      u.RateLimit.Limit,
+		Interval:   u.RateLimit.Interval,
+		// Only update rate limit TTL if the actor has been updated recently.
+		UpdateRateLimitTTL: u.LastUpdated != nil && time.Since(*u.LastUpdated) < 5*time.Minute,
```
Comment (eseliger) [inline enterprise/cmd/llm-proxy/internal/actor/actor.go:92]: should we make this interval configurable?

**Review by eseliger — COMMENTED** [review]
Looked over it briefly, the direction sounds good to me :)

File: enterprise/cmd/llm-proxy/shared/main.go [hunk 2]
```diff
@@ -175,27 +156,39 @@ func authenticate(logger log.Logger, cache httpcache.Cache, dotComClient graphql
...
-					errors.Errorf("failed to check access token: %s", err),
-				)
-				return
-			}
-
-			subscription = actor.NewSubscriptionFromDotCom(resp.Dotcom.ProductSubscriptionByAccessToken.ProductSubscriptionState)
-			saveSubscriptionToCache(token, subscription)
+		act, err := productSubscriptions.Get(r.Context(), token)
+		if err != nil {
+			responseJSONError(logger, w, http.StatusUnauthorized, err)
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/cmd/llm-proxy/shared/main.go:161]: No, I think right now the query returns an error if the access token is not associated with any subscription

[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/cmd/llm-proxy/shared/main.go:161]: it's not ideal, we should check for different error messages later

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/cmd/llm-proxy/shared/main.go:161]: could we separate 500 and 403 then? Doesn't have to be in this PR.

[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/llm-proxy/shared/main.go:161]: When I was writing the auth middleware PR, I wished the GQL message can be "typed" for "token not exist" or "something just wronng".

File: enterprise/cmd/llm-proxy/internal/actor/actor.go (hunk 1 above)
Comment (bobheadxi) [reply enterprise/cmd/llm-proxy/internal/actor/actor.go:92]: this is a difficult one to name

Comment (bobheadxi) [reply enterprise/cmd/llm-proxy/internal/actor/actor.go:92]: `LLM_PROXY_UPDATE_RATE_LIMIT_TTL_IF_LAST_UPDATED_WITHIN_THRESHOLD_MINUTES`

File: enterprise/cmd/llm-proxy/shared/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/cmd/llm-proxy/shared/main.go:161]: It is possible to add error codes in GraphQL!

File: enterprise/cmd/llm-proxy/internal/actor/actor.go (hunk 1 above)
Comment (unknwon) [reply enterprise/cmd/llm-proxy/internal/actor/actor.go:92]: LOL, time to take in structure config file 😂

File: enterprise/cmd/llm-proxy/shared/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/cmd/llm-proxy/shared/main.go:161]: yeah, I think we should check for the message at least and return more granular responses, @unknwon do you want to look at adding some easy-to-check error responses?

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/cmd/llm-proxy/shared/main.go:161]: https://sourcegraph.com/-/editor?remote_url=github.com%2Fsourcegraph%2Fsourcegraph&branch=es%2Fsubscriptions-ui&file=enterprise%2Fcmd%2Ffrontend%2Finternal%2Fbatches%2Fresolvers%2Ferrors.go&editor=VSCode&version=2.2.16&start_row=4&start_col=0&end_row=15&end_col=0

**Review by eseliger — APPROVED** [review]

File: enterprise/cmd/llm-proxy/shared/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/llm-proxy/shared/main.go:162]: Will take a look

**Pull request merged**